pkg/chains/storage/grafeas/grafeas.go

```diff
@@ -186,117 +184,191 @@ func (b *Backend) Type() string {
+		if err != nil {
+			return nil, err
+		}
+		occs = append(occs, occ)
+		return occs, nil
 	}
 
-	return nil
+	// create Occurrence_Build for TaskRun
+	for _, uri := range allURIs {
+		occ, err := b.createBuildOccurrence(ctx, tr, payload, signature, fmt.Sprint("https://", uri))
+		if err != nil {
+			return nil, err
+		}
+		occs = append(occs, occ)
+	}
+	return occs, nil
 }
 
-func (b *Backend) createOccurrenceRequest(tr *v1beta1.TaskRun, payload []byte, signature string, opts config.StorageOpts) (*pb.CreateOccurrenceRequest, error) {
+func (b *Backend) createAttestationOccurrence(ctx context.Context, tr *v1beta1.TaskRun, payload []byte, signature string, uri string) (*pb.Occurrence, error) {
 	occurrenceDetails := &pb.Occurrence_Attestation{
-		Attestation: &attestationpb.Details{
-			Attestation: &attestationpb.Attestation{
-				Signature: &attestationpb.Attestation_GenericSignedAttestation{
-					GenericSignedAttestation: &attestationpb.GenericSignedAttestation{
-						ContentType:       b.getContentType(opts),
-						SerializedPayload: payload,
-						Signatures: []*commonpb.Signature{
-							{
-								Signature:   []byte(signature),
-								PublicKeyId: b.cfg.Signers.KMS.KMSRef, // TODO: currently we only support storing kms keyID, will add other keys' ids later i.e. k8s secret, fulcio
-							},
-						},
-					},
+		Attestation: &pb.AttestationOccurrence{
+			SerializedPayload: payload,
+			Signatures: []*pb.Signature{
+				{
+					Signature: []byte(signature),
+					// TODO: currently we only support storing KMS keyID, may add other keys' ids later i.e. k8s secret, fulcio
+					PublicKeyId: b.cfg.Signers.KMS.KMSRef,
 				},
 			},
 		},
 	}
-
-	envelope := &commonpb.Envelope{
+	envelope := &pb.Envelope{
 		Payload:     payload,
-		PayloadType: "in-toto attestations containing a slsa.dev/provenance predicate",
-		Signatures: []*commonpb.EnvelopeSignature{
+		PayloadType: simplesigningPayloadType,
+		Signatures: []*pb.EnvelopeSignature{
 			{
-				Sig:   []byte(signature),
-				Keyid: b.cfg.Signers.KMS.KMSRef, // TODO: currently we only support storing kms keyID, will add other keys' ids later i.e. k8s secret, fulcio
+				Sig: []byte(signature),
+				// TODO: currently we only support storing KMS keyID, may add other keys' ids later i.e. k8s secret, fulcio
+				Keyid: b.cfg.Signers.KMS.KMSRef,
 			},
 		},
 	}
 
-	uri, err := b.getResourceURI(tr, opts)
-	if err != nil {
+	return b.client.CreateOccurrence(ctx,
+		&pb.CreateOccurrenceRequest{
+			Parent: b.getProjectPath(),
+			Occurrence: &pb.Occurrence{
+				ResourceUri: uri,
+				NoteName:    b.getAttestationNotePath(),
+				Details:     occurrenceDetails,
+				Envelope:    envelope,
+			},
+		},
+	)
+}
+
+func (b *Backend) createBuildOccurrence(ctx context.Context, tr *v1beta1.TaskRun, payload []byte, signature string, uri string) (*pb.Occurrence, error) {
+	in := intoto.ProvenanceStatement{}
+	if err := json.Unmarshal(payload, &in); err != nil {
```

Comment:
It'd be interesting to just pass the intoto struct reference around rather than generating the intoto statement -> serializing it -> de-serializing it -> converting it to grafeas.

Might be worth looking into to simplify things (we can also do this in another PR if needed)

Reply:
Yeah, agree that for grafeas backend, there are extra steps as you just described.

If we want to simplify that, we might need to change the function `StorePayload`'s signature. But the sacrifice is that other backend options have to serialize when `StorePayload` is called, right?

Reply:
Yeah... I think that's okay though? Seems reasonable to keep it as it's full spec and only serialize when we actually make the request. Fine for this now, but let's look into cleaning this up in another PR.